src/compiler/checker.rs

```diff
@@ -611,11 +623,11 @@ impl TypeChecker {
         }
     }
 
-    fn type_to_string(&self, type_: &Type) -> String {
+    fn type_to_string(&self, type_: Rc<Type>) -> String {
         let writer = Rc::new(RefCell::new(create_text_writer("")));
-        let type_node = self
-            .node_builder
-            .type_to_type_node(type_, Some(&*(*writer).borrow()));
+        let type_node =
+            self.node_builder
+                .type_to_type_node(self, type_, Some(&*(*writer).borrow()));
```

Comment:
Most of the `NodeBuilder` methods need to get a `&TypeChecker` reference injected to be able to call back to type-checker methods